crypto/provider_core.c

```diff
@@ -869,6 +874,13 @@ const OSSL_ALGORITHM *ossl_provider_query_operation(const OSSL_PROVIDER *prov,
     return prov->query_operation(prov->provctx, operation_id, no_cache);
 }
 
+void ossl_provider_query_free(const OSSL_PROVIDER *prov,
+                              const OSSL_ALGORITHM *algs)
```

Comment:
I'm unhappy with this interface, as it implies that the caller has ownership of the OSSL_ALGORITHM array, which is a huge change...

I would very much prefer if it mimicked the query interface, i.e. this:

``` C
int ossl_provider_unquery_operation(const OSSL_PROVIDER *prov,
                                    int operation_id)
```

Reply:
I don't agree.  This means every provider needs to keep a table of operations -> algorithms and know about thread IDs.  That's not easy.

The query returns the OSSL_ALGORITHM pointer, that's what needs to be freed.  Think of it as malloc() and free().

Adding the operation_id is a possibility but I don't think it is helpful.

Reply:
> I don't agree. This means every provider needs to keep a table of operations -> algorithms and know about thread IDs. That's not easy.

Er, huh?  The query operation needs to keep track of that already, so how would the unquery not be capable of doing the same thing?

> The query returns the OSSL_ALGORITHM pointer, that's what needs to be freed. Think of it as malloc() and free().

And that's exactly the point, the query is *not* designed to be a malloc.  It's return value being `const` should be a dead give-away!

What you're implying here is that `query_operation` would be allowed to return multiple copies of the same OSSL_ALGORITHM table...  and at the point, I'm not at all sure how `no_cache` is meaningful any more, we could simply assume that the providers that don't want us to remember the OSSL_ALGORITHM table could simply create a new one every time `query_operation` is called...  but then they must also make sure each copy sticks around since the caller may choose to keep a copy, which is the contrary of the idea with `no_cache`.

This is a huge change in OSSL_ALGORITHM handling, like I said.

Reply:
It may be that my explanation in https://github.com/openssl/openssl/issues/11771#issuecomment-698799382 is better

Reply:
Two threads: A and B.  A calls `query_operation` and gets a result.  Things change.  B calls `query_operation` and gets a different result.  One of them calls the free function passing the `operation_id`.  What gets freed?

It is far easier on the provider side to call `OPENSSL_free` on the passed `OSSL_ALGORITHM` pointer than it is to figure out what is no longer necessary.  It makes no difference on the caller's side which is passed back.

As I wrote, I'm happy to include the `operation_id` in addition to the `OSSL_ALGORITHM` pointer.  This makes no difference to the caller and might be useful to the provider.

Reply:
With the query/unquery model, the provider could have a per-operation_id refcount that gets ticked up by a query call, and ticked down by an unquery call.  That fulfills the same objective without having to allocate and deallocate memory all over the place.

Reply:
So query/unquery instead of query/free?  I fear we're arguing about a triviality.

Reply:
It's a little more than a triviality, I believe it sends a message.  With something that implies a malloc()/free() line of thought, provider authors must deal with memory management in a whole new way.  I can easily imagine that an author might come to think that they must somehow allocate a new OSSL_ALGORITHM... exactly when? ... and what's supposed to happen with that pointer, are they relinquishing ownership (I suspect not, but the question will still be there)?  Is that an idea that we will entertain in the future?  Will providers have to deal with the thought that there may be multiple OSSL_ALGORITHMs for the same operation ID floating around in memory somewhere?

No, I don't think this is a triviality.  Or well, it can be argued that the names are...  but the argument being passed as far from a triviality.

Reply:
Providers can't relinquish ownership of memory they allocated -- by design.

A non-caching provider _will_ have to deal with multiple OSSL_ALGORITHMs for the same operation.  Plug in a different card, potentially a different array.

I've added the _operation_id_ to the argument list.  I've kept the array returned from the query function as an argument.  In the case of multiple OSSL_ALGORITHMs for the same operation, this is the only distinguishing factor.  We don't need it now but we will need it eventually.

Reply:
> A non-caching provider _will_ have to deal with multiple OSSL_ALGORITHMs for the same operation. Plug in a different card, potentially a different array.

Would that be *multiple* cards plugged in to the same device simultaneously?

This is news to me...  if that was implied in our early design meetings (I do recall that there was a discussion around PKCS#11, smartcards, that sort of stuff), I didn't catch this aspect.

Reply:
Unplug, replug.  A fetched algorithm for the first card could still be around -- the provider will need to accept calls and return errors.

Reply:
> Unplug, replug. A fetched algorithm for the first card could still be around -- the provider will need to accept calls and return errors.

Yes?  So you have the method structure around.  That doesn't impact on what `OSSL_ALGORITHM` are expected to be around.

Reply:
It impacts the provider knowing if an OSSL_ALGORITHM is being used or not.

Reply:
> It impacts the provider knowing if an OSSL_ALGORITHM is being used or not.

Doesn't the a query/unquery pair with the semantics I'd like to go for do that?

Reply:
Ah, I think I see the misconception here.  Currently, none of the method constructors save the OSSL_ALGORITHM pointer *at all*, and if the provider sets `*no_store` to 1, they are not allowed to.  So the OSSL_ALGORITHM per se is only used in the relatively brief moment where there is an ongoing fetch happening, and the unquery is the signal back to the provider that "done now!".

They idea, though, is that at any one time, the query function only has one OSSL_ALGORITHM to return and deal with, and I can't see that you have said anything different.

Reply:
In a multi-threaded environment, the _only one OSSL_ALGORITHM_ assumption is broken.  A thread  could receive a different OSSL_AGORITHM return while another thread is still using theirs.

Reply:
You know what, I give in.